modules/ztp-pre-staging-downloading-artifacts.adoc

```diff
@@ -0,0 +1,416 @@
++
+[source,yaml]
+----
+apiVersion: mirror.openshift.io/v1alpha2
+kind: ImageSetConfiguration
+mirror:
+  platform:
+---
+  operators:
+    - catalog: eko4.cloud.lab.eng.bos.redhat.com:8443/redhat/redhat-operator-index:v4.11 <1>
+      packages:
+        - name: advanced-cluster-management
+          channels:
+             - name: 'release-2.6'
+             - name: 'release-2.5'
+               minVersion: 2.5.4
+               maxVersion: 2.5.4
+        - name: multicluster-engine
+          channels:
+             - name: 'stable-2.1'
+             - name: 'stable-2.0'
+               minVersion: 2.0.4
+               maxVersion: 2.0.4
+        - name: local-storage-operator
+          channels:
+            - name: 'stable'
+        - name: ptp-operator
+          channels:
+            - name: 'stable'
+        - name: sriov-network-operator
+          channels:
+            - name: 'stable'
+        - name: cluster-logging
+          channels:
+            - name: 'stable'
+    - catalog: eko4.cloud.lab.eng.bos.redhat.com:8443/redhat/certified-operator-index:v4.11 <1>
+----
+<1> The catalog source is changed from an official Red Hat registry.
+// On line 315 and 341 we have the same field with the same value, equal identation. Is that intentional? If yes, what does each one do?
+
+When you download images by using a local or disconnected registry, you have to first add certificates for the registries that you want to pull content from.
+
+. To avoid any errors, copy the registry certificate into your server:
++
+[source,terminal]
+----
+# cp /tmp/eko4-ca.crt /etc/pki/ca-trust/source/anchors/.
+----
+
+. Then, update the certificates trust store:
++
+[source,terminal]
+----
+# update-ca-trust
+----
+
+. Mount the host `/etc/pki` folder into the factory-cli image:
++
+[source,terminal]
+----
+# podman run -v /mnt:/mnt -v /root/.docker:/root/.docker -v /etc/pki:/etc/pki --privileged -it --rm quay.io/user/telco-ran-tools:acm-versioning -- \
+    factory-precaching-cli download -r 4.11.5 -f /mnt \
+    --ai-img registry.redhat.io/multicluster-engine/assisted-installer-agent-rhel8@sha256:da1753f9fcb9e229d0a68de03fac90d15023e647a8db531ae489eb93845d5306 \
+    --ai-img registry.redhat.io/multicluster-engine/assisted-installer-reporter-rhel8@sha256:e8d6b78248352b1a8e05a22308185a468d4a139682d997a7f968b329abbc02cd \
+    --ai-img registry.redhat.io/multicluster-engine/assisted-installer-rhel8@sha256:33abd6e21cfdc36dd4337fa6f3c3442d33fc3f976471614dca5b8ef749e7a027 \
+    --img quay.io/user/troubleshoot \
+    --du-profile --acm-hub 2.5.4 -s \
+    --skip-imageset
+----
+
+////
+Could we have an example where this goes through? I feel like it makes more sense to include adding the certs before the user encounters the error. Let's point it out and add the steps here. I can include a troubleshooting section in case they skip adding the certs
+
+Also, if we do it this way, the remaining step might not be required, if I understand it correctly.
```

Comment:
If you mean the `--skip-imageset` step, this is paired with the `--generate-imageset` option to split the procedure.

@alosadagrande, correct me if I'm wrong, but I believe the intent here is that the user must add whatever certs are needed for accessing the disconnected registries. It's not that this is being done if an error occurs - it's a required step for accessing the disconnected registry. So the steps at a high level for using a disconnected registry would be:

1. Run `factory-precaching-cli download --generate-imageset`
2. Update generated imageset.yaml to replace catalog servers/references with disconnected registry server
3. Install required certificate for accessing registry server
4. Run `factory-precaching-cli download --skip-imageset` to perform the image downloads using the modified imageset.yaml

So step 3, installing the certs, could be done at any point before step 4... If you don't do this step, you'll get failures during that final step, as the image download requests will get rejected.

Reply:
I meant an example where you added the certs and you successfully downloaded the images. However, looking at it now, that might not be needed.

On the other half of my question/suggestion: The structure in the upstream docs was my main issue because it showed extra steps that could be avoided. Even more simplified, it was:
1. Generate ImageSetConfig.
2. Modify ImageSetConfig.
3. Try downloading the images.
4. Get error.
5. Resolve the issue by adding the certs.
6. Download the images.

So to avoid seeing the error at all, I restructured it to:
1. Generate ImageSetConfig.
2. Modify ImageSetConfig.
3. Add certs.
4. Download the images.

I think we're on the same page, but I wanted to clear up my initial q, just in case :)

Reply:
Exactly. If you're going to pull images from a disconnected registry you need to include the proper certificates. In this case, since we are using a container image we need to mount the host's PKI folder.

Notice that this is the case in disconnected environments, and concretely, when you replace the catalog source with a disconnected one. In that case, you need to pull the catalog source (from a non-RH registry) and that catalog source will tell you the location (non-RH registry) of the operators required to be pre-cached.